[android] update AOT profile for .NET 8.0.0-rc.1.23419.3

jonathanpeppers opened the pull request:
I recorded this profile using newer builds of .NET 8 RC 1.

An average of 10 runs of a dotnet new maui app on a Pixel 5, the new profile helps quite a bit:

    Before:
    Average(ms): 693.2
    Std Err(ms): 5.25314720482451
    Std Dev(ms): 16.6119100513925
    After:
    Average(ms): 607.8
    Std Err(ms): 2.94316534061922
    Std Dev(ms): 9.307106006822

Things appear to be even slower than last time in 3f221532. 😢

jonathanpeppers committed:
[android] update AOT profile for .NET 8.0.0-rc.1.23419.3

I recorded this profile using newer builds of .NET 8 RC 1.

An average of 10 runs of a dotnet new maui app on a Pixel 5, the new
profile helps quite a bit:

    Before:
    Average(ms): 693.2
    Std Err(ms): 5.25314720482451
    Std Dev(ms): 16.6119100513925
    After:
    Average(ms): 607.8
    Std Err(ms): 2.94316534061922
    Std Dev(ms): 9.307106006822

Things appear to be even slower than last time in 3f221532. 😢

jonathanpeppers commented:
If you look here, we are really close to on-par with .NET 7:

https://aka.ms/dotnetperfstatus

![image](https://github.com/dotnet/maui/assets/840039/ff6135f5-9b1b-4fd0-a41a-cbbe4a1672cd)

The giant orange mountain was before https://github.com/dotnet/maui/commit/3f22153219859e79d590b37ea0cc15abe6ba6186 was merged.

So, I wonder what will happen after this is merged.

mattleibow commented on src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/maui.aotprofile.txt:
```diff
@@ -323,45 +261,31 @@ Methods:
 	bool Microsoft.Maui.Controls.Layout:Microsoft.Maui.ILayout.get_ClipsToBounds ()
 	bool Microsoft.Maui.Controls.LayoutOptions:Equals (object)
 	bool Microsoft.Maui.Controls.MergedStyle:<OnClassStyleChanged>b__29_1 (Microsoft.Maui.Controls.Style)
-	bool Microsoft.Maui.Controls.NavigationPage:<FindMyToolbar>b__98_0 (Microsoft.Maui.Controls.Element)
-	bool Microsoft.Maui.Controls.NavigationPage:GetHasBackButton (Microsoft.Maui.Controls.Page)
-	bool Microsoft.Maui.Controls.NavigationPage:GetHasNavigationBar (Microsoft.Maui.Controls.BindableObject)
-	bool Microsoft.Maui.Controls.NavigationPage/<>c:<FindMyToolbar>b__98_1 (Microsoft.Maui.Controls.Element)
```
A whole bunch of nav stuff is missing, is this expected?

mattleibow commented on src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/maui.aotprofile.txt:
```diff
@@ -90,14 +85,11 @@ Methods:
 	Android.Runtime.IdentityHashTargets& System.Collections.Generic.Dictionary`2<System.IntPtr, Android.Runtime.IdentityHashTargets>:FindValue (intptr)
 	Android.Runtime.JavaArray`1<string> Android.Runtime.JavaArray`1<System.String>:FromJniHandle (intptr,Android.Runtime.JniHandleOwnership)
 	Android.Util.DisplayMetrics Android.Content.Res.Resources:get_DisplayMetrics ()
-	Android.Views.Display Android.Views.IWindowManagerInvoker:get_DefaultDisplay ()
-	Android.Views.Display Microsoft.Maui.Devices.DeviceDisplayImplementation:GetDefaultDisplay ()
```
And we don;t need the display density? I think this is always needed once...

mattleibow commented on src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/maui.aotprofile.txt:
```diff
@@ -1836,20 +1457,14 @@ Methods:
 	Microsoft.Maui.Controls.BindingBase Microsoft.Maui.Controls.Xaml.BindingExtension:Microsoft.Maui.Controls.Xaml.IMarkupExtension<Microsoft.Maui.Controls.BindingBase>.ProvideValue (System.IServiceProvider)
 	Microsoft.Maui.Controls.Brush Microsoft.Maui.Controls.Brush:get_Default ()
 	Microsoft.Maui.Controls.Brush Microsoft.Maui.Controls.Brush:op_Implicit (Microsoft.Maui.Graphics.Color)
-	Microsoft.Maui.Controls.Brush Microsoft.Maui.Controls.NavigationPage:get_BarBackground ()
```
Looks like the nav bar is not needed at all?

jonathanpeppers replied:
Ok, I think the app is doing a web request (GET) that is timing out:

https://httpstat.us/200

This is not completing for me. Let me investigate and record it again.

Maybe we will be in OK shape if I fix this!

jonathanpeppers replied:
Going to record again with: https://httpbin.org/status/200

mattleibow reviewed (commented):
This looks a bit interesting... Did the android app actually load?

jonathanpeppers committed:
Use https://httpbin.org/status/200

https://httpstat.us/200 appears to be down right now.

The new profile is slightly better:

    Average(ms): 594.6
    Std Err(ms): 3.06304133537604
    Std Dev(ms): 9.68618718703197

But still appears to be a smidge worse than 3f221532.

jonathanpeppers commented:
The new profile (with https://httpbin.org/status/200) is slightly better:

    Average(ms): 594.6
    Std Err(ms): 3.06304133537604
    Std Dev(ms): 9.68618718703197

jonathanpeppers commented on src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/maui.aotprofile.txt:
```diff
@@ -3371,13 +3102,16 @@ Methods:
 	System.Reflection.CustomAttributeTypedArgument[] System.Reflection.RuntimeCustomAttributeData:UnboxValues (object[])
-	System.Reflection.FieldAttributes System.Reflection.RuntimeFieldInfo:get_Attributes ()
-	System.Reflection.FieldInfo System.Net.CookieParser:get_IsQuotedDomainField ()
+	System.Reflection.Emit.ILGenerator System.Reflection.Emit.DynamicMethod:GetILGenerator (int)
+	System.Reflection.Emit.Label System.Reflection.Emit.RuntimeILGenerator:BeginExceptionBlock ()
+	System.Reflection.Emit.Label System.Reflection.Emit.RuntimeILGenerator:DefineLabel ()
+	System.Reflection.Emit.LocalBuilder System.Reflection.Emit.ILGenerator:DeclareLocal (System.Type)
+	System.Reflection.Emit.LocalBuilder System.Reflection.Emit.RuntimeILGenerator:DeclareLocal (System.Type,bool)
+	System.Reflection.Emit.RuntimeILGenerator System.Reflection.Emit.DynamicMethod:GetILGeneratorInternal (int)
```
These are new; going to figure out where they are coming from.

jonathanpeppers replied:
I'm going to try to reproduce this in a `dotnet new android` app, but what happens is these end up in the profile:

* https://github.com/dotnet/runtime/blob/819e288da8689a5aefe9f201865b0f44f00537d4/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ILEmit/ILEmitResolverBuilder.cs#L63
* https://github.com/dotnet/runtime/blob/819e288da8689a5aefe9f201865b0f44f00537d4/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ILEmit/ILEmitResolverBuilder.cs#L351

But they are never called when I look at `dotnet-trace`.

What's even weirder though, is the runtime loads the code for them:
```
08-23 16:30:54.710  3690  3752 D Mono    : AOT: FOUND method Microsoft.Extensions.DependencyInjection.ServiceLookup.ILEmitResolverBuilder:GenerateMethodBody (Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite,System.Reflection.Emit.ILGenerator) [0x74cb72f3c0 - 0x74cb7301d0 0x74cb73701f]
...
08-23 16:30:54.712  3690  3752 D Mono    : AOT: FOUND method System.Reflection.Emit.RuntimeILGenerator:BeginExceptionBlock () [0x74d4632670 - 0x74d4632950 0x74d46a4c54
```

So, the usage of a delegate, seems like it causes the runtime to "preload" AOT images for code that is never run? I think a smaller repro might help explain what is going on.

mattleibow reviewed (changes requested):
Just marking as needs changes to prevent any merging while @jonathanpeppers investigates the reflection scare!

jonathanpeppers commented:
@mattleibow no we should merge this. It looks like the change would be in dotnet/runtime anyway and it closed for RC 1. This is better than nothing for sure. About 100ms savings.
